uniffi/src/lib.rs

```diff
@@ -143,6 +143,14 @@ macro_rules! assert_compatible_version {
 /// implementations generated from your component UDL via the `uniffi-bindgen scaffolding` command.
 
 pub unsafe trait ViaFfi: Sized {
+    /// The type used in rust code.
+    ///
+    /// For primitive / standard types, we implement ViaFfi on the type itself and RustType=Self.
+    /// For user-defined types we create a unit struct and implement it there.  This sidesteps
+    /// Rust's orphan rules with types from external crates, since we're not implementing ViaFfi on
+    /// the external type directly.
+    type RustType;
```

Comment:
> I think it should support the adapter pattern pretty well

If you want to lean in to the design-patterns language here, `FFIAdaptor` seems like it would be a fine name for this trait.

Reply:
Also, would it make sense to default `RustType` to `Self` here, i.e. `type RustType = Self`? This might help encourage the right mental model of "well usually we implement this on the type itself, but sometimes we implement it on a helper struct in special cases".

Reply:
> , i.e. `type RustType = Self`?

IIRC, I tried that in one of my earlier experiments, and was greeted with:

`error[E0658]: associated type defaults are unstable` :(

Reply:
The `ViaFfi` impls in `uniffi` all have `RustType=Self`.  The ones generated in the bindings don't.  Here's an example from the external lib scaffolding:
```
  struct UniffiTypeRecordCrateOneType;
  
  #[doc(hidden)]
  impl uniffi::RustBufferViaFfi for UniffiTypeRecordCrateOneType {
      type RustType = CrateOneType;
   ... 
}
```

I think changing the name makes sense and I'm open to anything, except something with `Adapter` in it.  I'm hoping to write an adapter trait which will help with things like adapting the String impl to output a `serde_json::Value`.  I really don't want to name that trait `FFIAdapterAdapter` :)

Using generics seems interesting.  I think in the past it would make it harder to implement `call_with_result`, but maybe not anymore.  I'm pretty unclear about the pros and cons here.  I will say that I don't think we'll need 1 struct to implement 2 conversions.  The nice thing about this system is we can always just create a new unit struct to do whatever conversion we want.

And yeah, I would love to set a default for `RustType`, but I got the unstable warning.